File: 2018/day03/part1.py
# #1 @ 469,741: 22x26
import re

pattern = re.compile("#(\d+) @ (\d+),(\d+): (\d+)x(\d+)")
def parse(line):
    m = pattern.match(line)
    idx, left, top, width, height = map(int, m.groups())
    return {'id': idx, 'left': left, 'top': top, 'width': width, 'height': height}


grid = [[0 for _ in range(1000)] for _ in range(1000)]
def modify_grid(m):
    for x in range(m['left'], m['left'] + m['width'] ):
        for y in range(m['top'], m['top'] + m['height']):
            grid[x][y] += 1

def is_claim_intact(m):
    for x in range(m['left'], m['left'] + m['width'] ):
        for y in range(m['top'], m['top'] + m['height']):
            if grid[x][y] != 1:
                return False
    return True

with open('input.txt', 'r') as f:
    measurements = [parse(line) for line in f]
    for m in measurements:
        modify_grid(m)

    count = 0
    for row in grid:
        for item in row:
            if item > 1:
                count += 1
    print("Part 1 answer is: {}".format(count))

    for m in measurements:
        if is_claim_intact(m):
            print("Part 2 answer is: {}".format(m['id']))

File: 2018/day03/part1.clj
; #1 @ 469,741: 22x26
(require '[clojure.string :as str])

(defn parse-int [v] (Integer/parseInt v))

(defn get-input []
  (str/split (slurp "input.txt") #"\n"))

(defn parse-line [line]
  (let [
        [idx, l, t, w, h] (map parse-int (drop 1 (re-find #"#(\d+) @ (\d+),(\d+): (\d+)x(\d+)" line)))]
    {:id idx, :left l, :top t, :width w, :height h}))

(defn get_grid []
  (reduce
    (fn [grd, c]
      (for [x (range (get c :left) (+ (get c :left) (get c :width)))
            y (range (get c :top) (+ (get c :top) (get c :height)))]
        (assoc-in grd [x y] (inc (get-in grd [x y])))))
    (vec (repeat 1001 (vec (repeat 1001 0))))
    (map parse-line (get-input))
  ))

(get_grid)

(println
  (let [filled_grid (get_grid)]
    (reduce
      (fn [acc [x y]]
        (do (println acc x y)
        (if (< 1 (get-in filled_grid [x y]))
          (inc acc)
          acc)))
      0
      (for [x (range 1000)
            y (range 1000)] [x y]))))

